src/Compilers/CSharp/Portable/Symbols/ExtraAnnotations.cs

```diff
@@ -0,0 +1,202 @@
+            {
+                return null;
+            }
+
+            var pooledBuilder = PooledStringBuilder.GetInstance();
+
+            StringBuilder builder = pooledBuilder.Builder;
+            Add(method.ReturnType.TypeSymbol, builder);
+            builder.Append(' ');
+
+            Add(containingType, builder);
+            builder.Append('.');
+
+            builder.Append(method.Name);
+            builder.Append('(');
+
+            var parameterTypes = method.ParameterTypes;
+            for (int i = 0; i < parameterTypes.Length; i++)
+            {
+                Add(parameterTypes[i].TypeSymbol, builder);
+                if (i < parameterTypes.Length - 1)
+                {
+                    builder.Append(", ");
+                }
+            }
+
+            builder.Append(')');
+            return pooledBuilder.ToStringAndFree();
+
+            // PROTOTYPE(NullableReferenceTypes): Many cases are not yet handled
+            // generic type args
+            // ref kind
+            // 'this'
+            // static vs. instance
+            // use assembly qualified name format (used in metadata) rather than symbol display?
+        }
+
+        private static ImmutableArray<ImmutableArray<bool>> Array(params ImmutableArray<bool>[] values)
+            => values.ToImmutableArray();
+
+        private static ImmutableArray<AttributeAnnotations> Array(params AttributeAnnotations[] values)
+            => values.ToImmutableArray();
+
+        private static ImmutableArray<bool> Nullable(params bool[] values)
+        {
+            Debug.Assert(values.Length > 0);
+            return values.ToImmutableArray();
+        }
+
+        internal static ImmutableArray<ImmutableArray<bool>> GetExtraAnnotations(string key)
+        {
+            if (key is null)
+            {
+                return default;
+            }
+
+            if (!Annotations.TryGetValue(key, out var flags))
+            {
+                return default;
+            }
+
+            return flags;
+        }
+
+        private static void Add(TypeSymbol type, StringBuilder builder)
+            => builder.Append(
+                type.ToDisplayString(
+                    SymbolDisplayFormat.CSharpErrorMessageFormat
+                        .RemoveMiscellaneousOptions(SymbolDisplayMiscellaneousOptions.UseSpecialTypes)
+                        .AddMiscellaneousOptions(SymbolDisplayMiscellaneousOptions.IncludeNullableReferenceTypeModifier)
+                        // displaying tuple syntax causes to load the members of ValueTuple, which can cause a cycle, so we use long-hand format instead
+                        .WithCompilerInternalOptions(SymbolDisplayCompilerInternalOptions.UseValueTuple)));
+
+        /// <summary>
+        /// index 0 is used for return type
+        /// other parameters follow
+        /// </summary>
+        internal static (bool hasAny, AttributeAnnotations annotations) GetExtraAttributes(string key, int parameterIndex)
```

Comment:
>bool hasAny [](start = 25, length = 11)

I am not sure what extra information do we get from ```hasAny``` flag. Would it be sufficient to simply check ````AttributeAnnotations``` for 0?

Reply:
Perhaps a clear comment might help

---
In reply to: [187938913](https://github.com/dotnet/roslyn/pull/26656#discussion_r187938913) [](ancestors = 187938913)

Reply:
Added in follow-up PR https://github.com/dotnet/roslyn/pull/26828/files#diff-d02e2d6d70e84697ca87520824236d04R169